luna/gateware/usb/usb2/descriptor_single_memory.py

```diff
@@ -0,0 +1,438 @@
+
+from nmigen                                  import *
+from usb_protocol.emitters.descriptors       import DeviceDescriptorCollection
+from usb_protocol.types.descriptors.standard import StandardDescriptorNumbers
+
+from ..stream                          import USBInStreamInterface
+from ...stream.generator               import ConstantStreamGenerator
+from ...test                           import LunaUSBGatewareTestCase, usb_domain_test_case
+
+
+class GetDescriptorHandlerSingleMemory(Elaboratable):
+    """ Gateware that handles responding to GetDescriptor requests.
+
+    Currently does not support descriptors in multiple languages.
+
+    I/O port:
+        I: value[16]      -- The value field associated with the Get Descriptor request.
+                             Contains the descriptor type and index.
+        I: length[16]     -- The length field associated with the Get Descriptor request.
+                             Determines the maximum amount allowed in a response.
+
+        I: start          -- Strobe that indicates when a descriptor should be transmitted.
+        I: start_position -- Specifies the starting position of the descriptor data to be transmitted.
+
+        *: tx             -- The USBInStreamInterface that streams our descriptor data.
+        O: stall          -- Pulsed if a STALL handshake should be generated, instead of a response.
+    """
+
+    def __init__(self, descriptor_collection: DeviceDescriptorCollection, max_packet_length=64, domain="usb"):
+        """
+        Parameters
+        ----------
+        descriptor_collection: DeviceDescriptorCollection
+            The DeviceDescriptorCollection containing the descriptors to use for this device.
+        max_packet_length: int
+            Maximum packet length.
+        domain: string
+            The clock domain this generator should belong to. Defaults to 'usb'.
+        """
+
+        self._descriptors = descriptor_collection
+        self._max_packet_length = max_packet_length
+        self._domain           = domain
+
+        #
+        # I/O port
+        #
+        self.value          = Signal(16)
+        self.length         = Signal(16)
+
+        self.start          = Signal()
+        self.start_position = Signal(11)
+
+        self.tx             = USBInStreamInterface()
+        self.stall          = Signal()
+
+    def generate_rom_content(self):
+        """ Generates the ROM content.
+
+        Memory layout
+        -------------
+
+        All data is aligned on 4 bytes
+
+        Type offsets and number of entries
+        ----------------------------------
+
+        Each descriptor type starting from 0 until maximum used type number has
+        an entry consisting of number of indexes for this type number (2 bytes)
+        and address of first index (2 bytes).
+        Invalid entries have a value of 0x0000xxxx (0 entries).
+
+        Example:
+        0000  0xFFFF
+        0002  0xFFFF
+        0004  Number of device indexes
```

Comment:
Correct latin plural would be 'indices'

Reply:
All sources I find (e.g. https://en.wiktionary.org/wiki/index) list `indexes` as well, often as their 'first' choice.